parlai/scripts/jit_export.py

```diff
@@ -0,0 +1,751 @@
+
+        # Format history vec given various logic
+        if self.text_truncate is not None:
+            if self.is_bart:
+                truncate_length = self.text_truncate - 2  # Start and end tokens
+            else:
+                truncate_length = self.text_truncate
+            if len(flattened_text_vec) > truncate_length:
+                flattened_text_vec = flattened_text_vec[-truncate_length:]
+        flattened_text_vec = torch.tensor(flattened_text_vec, dtype=torch.long)
+        if self.is_bart:
+            flattened_text_vec = torch.cat(
+                [
+                    torch.tensor([self.start_idx], dtype=torch.long),
+                    flattened_text_vec,
+                    torch.tensor([self.end_idx], dtype=torch.long),
+                ],
+                dim=0,
+            )
+
+        # Pass through the encoder and decoder to generate tokens
+        batch_text_vec = torch.unsqueeze(flattened_text_vec, dim=0)  # Add batch dim
+        encoder_states = self.encoder(batch_text_vec)
+        generations = self._get_initial_decoder_input(batch_text_vec)
+        # keep track of early stopping if all generations finish
+        seen_end = torch.zeros(
+            batch_text_vec.size(0), device=batch_text_vec.device, dtype=torch.bool
+        )
+        incr_state: Dict[str, torch.Tensor] = {}
+        for token_idx in range(max_len):
+            if token_idx == 0:
+                latent, incr_state = self.decoder_first_pass(
+                    generations, encoder_states
+                )
+            else:
+                latent, incr_state = self.decoder_later_pass(
+                    generations, encoder_states, incr_state
+                )
+            logits = self.partially_traced_model.output(latent[:, -1:, :])
+            _, preds = logits.max(dim=2)
+            incr_state = self.partially_traced_model.reorder_decoder_incremental_state(
+                incr_state,
+                torch.tensor([0], dtype=torch.long, device=batch_text_vec.device),
+            )
+            seen_end = seen_end + (preds == self.end_idx).squeeze(1)
+            generations = torch.cat([generations, preds], dim=1)
+            if torch.all(seen_end):
+                break
+
+        # Get the label from the generated tokens and update the history
+        if self.is_bart:
+            assert generations[0, 0].item() == self.end_idx
+            generations = generations[:, 1:]
+            # Hack: remove initial end token. I haven't found in the code where this is
+            # done, but it seems to happen early on during generation
+        generation_tokens: List[int] = generations[0].tolist()
+        label = self._v2t(generation_tokens)
+
+        return label
+
+
+@torch.jit.script
+class ScriptableGpt2BpeHelper(object):
+    """
+    Version of parlai.utils.bpe.Gpt2BpeHelper that can be TorchScripted.
+    """
+
+    @classmethod
+    def findall(cls, text: str) -> List[str]:
```

Comment:
where is this method coming from?

Reply:
I had to re-write regex for splitting tokens because you can't TorchScript it -__-